### PR TITLE
lib: Introduce Stanford CoreNLP wrapper

### PR DESCRIPTION
**Pull request by Morriar**
## Stanford Core NLP

Stanford CoreNLP provides a set of natural language analysis tools which can take
raw text input and give the base forms of words, their parts of speech, whether
they are names of companies, people, etc., normalize dates, times, and numeric
quantities, and mark up the structure of sentences in terms of phrases and word
dependencies, indicate which noun phrases refer to the same entities, indicate
sentiment, etc.

This wrapper needs the Stanford CoreNLP jars that run on Java 1.8+.

See http://nlp.stanford.edu/software/corenlp.shtml.

## Usage

~~~nit
var proc = new NLPProcessor("path/to/StanfordCoreNLP/jars")

var doc = proc.process("String to analyze")

for sentence in doc.sentences do
	for token in sentence.tokens do
		print "{token.lemma}: {token.pos}"
	end
end
~~~

### CODE REVIEW

File: lib/nlp/stanford.nit [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+	#       </sentence>
+	#     </sentences>
+	#   </document>
+	# </root>""".to_xml.as(XMLDocument)
+	#
+	# var document = new NLPDocument.from_xml(xml)
+	# assert document.sentences.length == 2
+	# assert document.sentences.first.tokens.first.word == "Stanford"
+	# assert document.sentences.last.tokens.first.word == "UQAM"
+	init from_xml(xml: XMLDocument) do
```
Comment (privat) [inline lib/nlp/stanford.nit]: Unclosed fence in nitunit

File: lib/nlp/vsm.nit [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+# NLPVector Space Model.
+#
+# The Vector Space Model (VSM) is used to compare natural language texts.
+# Texts are translated to multidimensionnal vectors then compared by cosine
+# similarity.
+module vsm
+
+# A multi-dimensional vector.
+class NLPVector
+	super HashMap[String, Int]
```
Comment (privat) [inline lib/nlp/vsm.nit]: Maybe a naive idea but why not dropping NLPVector and using Counter instead?

**Comment by privat** [comment]
+1 very cool.

File: lib/nlp/vsm.nit (hunk 2 above)
Comment (Morriar) [inline lib/nlp/vsm.nit]: Good idea, I will try with this

File: lib/nlp/README.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+dependencies, indicate which noun phrases refer to the same entities, indicate
+sentiment, etc.
+
+This wrapper needs the Stanford CoreNLP jars that run on Java 1.8+.
+
+See http://nlp.stanford.edu/software/corenlp.shtml.
+
+## Usage
+
+~~~raw
```
Comment (privat) [inline lib/nlp/README.md]: `nitish` instead of `raw` to have highlighting

File: lib/nlp/stanford.nit [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+	#       <CharacterOffsetEnd>19</CharacterOffsetEnd>
+	#       <POS>NNP</POS>
+	#     </token>
+	#   </tokens>
+	# </sentence>""".to_xml["sentence"].first.as(XMLStartTag)
+	#
+	# var sentence = new  NLPSentence.from_xml(xml)
+	# assert sentence.index == 1
+	# assert sentence.tokens.length == 2
+	init from_xml(xml: XMLStartTag) do
```
Comment (privat) [inline lib/nlp/stanford.nit]: unclosed nitunit fence

File: lib/nlp/stanford.nit [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+				tokens.add new NLPToken.from_xml(obj)
+			else
+				print "Warning: malformed xml, `tokens` is supposed to contain `token` tags"
+			end
+		end
+		init(index)
+	end
+end
+
+# Represent one word (or puncutation mark) in a	` NLPSentence`.
```
Comment (privat) [inline lib/nlp/stanford.nit]: weird tabulation

File: lib/nlp/vsm.nit [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		for term in terms do
+			dot += self.get_or_default(term, 0) * other.get_or_default(term, 0)
+		end
+
+		return dot.to_f / (self.norm * other.norm)
+	end
+
+	# The norm of the vector.
+	#
+	# `||x|| = (x1.pow + ... + xn.pow).sqrt`
```
Comment (privat) [inline lib/nlp/vsm.nit]: `x1.pow`? do you mean something like `x1.sqr` or `x1**2` or `x1.pow(2)`

**Comment by Morriar** [comment]
Reroll with @privat remarks.

**Comment by privat** [comment]
+1

File: lib/nlp/stanford.nit [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	var java_cp: String
+
+	# Process a string and return a new NLPDocument from this.
+	fun process(string: String): NLPDocument do
+		var tmp_file = ".nlp.in"
+		var file = new FileWriter.open(tmp_file)
+		file.write string
+		file.close
+		var doc = process_file(tmp_file)
+		sys.system "rm {tmp_file}"
```
Comment (xymus) [inline lib/nlp/stanford.nit]: `file_delete` could do this for you.

**Comment by xymus** [comment]
+1

What services would you like to access from the JAR archive? I could try to generate something with jwrapper if you are interested.

**Comment by R4PaSs** [comment]
:shipit:

**Comment by Morriar** [comment]
Reroll: fixed @xymus remarks.

@xymus Appart the main process methods I don't know. I'm just fooling around with the jar for the moment.

**Pull request merged**